## **Задание 1.**

**Выберите датасет, который имеет отношение к вашей области интересов или исследований.
Датасет должен содержать неструктурированные данные, требующие разметки для решения конкретной задачи, например, анализа настроений или распознавания именованных сущностей.**

- Рассмотрим на **kaggle** и загрузим датасет
    
    https://www.kaggle.com/datasets/arashnic/book-recommendation-dataset?select=Books.csv

### 1: Изучение данных
1. **Загрузка данных и первичный анализ**:
   - Загрузим данные из файлов CSV и изучим их структуру

In [100]:
import pandas as pd

In [101]:
# Загрузка данных
books = pd.read_csv('Books.csv')
ratings = pd.read_csv('Ratings.csv')
users = pd.read_csv('Users.csv')

C:\Users\user\AppData\Local\Temp\ipykernel_7560\2183490731.py:2: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv('Books.csv')


In [102]:
# Проверка типов данных
print(books.dtypes)

ISBN                   object
Book-Title             object
Book-Author            object
Year-Of-Publication    object
Publisher              object
Image-URL-S            object
Image-URL-M            object
Image-URL-L            object
dtype: object


In [103]:
# Проверка уникальных значений в проблемном столбце
print(books.iloc[:, 3].unique())

[2002 2001 1991 1999 2000 1993 1996 1988 2004 1998 1994 2003 1997 1983
 1979 1995 1982 1985 1992 1986 1978 1980 1952 1987 1990 1981 1989 1984 0
 1968 1961 1958 1974 1976 1971 1977 1975 1965 1941 1970 1962 1973 1972
 1960 1966 1920 1956 1959 1953 1951 1942 1963 1964 1969 1954 1950 1967
 2005 1957 1940 1937 1955 1946 1936 1930 2011 1925 1948 1943 1947 1945
 1923 2020 1939 1926 1938 2030 1911 1904 1949 1932 1928 1929 1927 1931
 1914 2050 1934 1910 1933 1902 1924 1921 1900 2038 2026 1944 1917 1901
 2010 1908 1906 1935 1806 2021 '2000' '1995' '1999' '2004' '2003' '1990'
 '1994' '1986' '1989' '2002' '1981' '1993' '1983' '1982' '1976' '1991'
 '1977' '1998' '1992' '1996' '0' '1997' '2001' '1974' '1968' '1987' '1984'
 '1988' '1963' '1956' '1970' '1985' '1978' '1973' '1980' '1979' '1975'
 '1969' '1961' '1965' '1939' '1958' '1950' '1953' '1966' '1971' '1959'
 '1972' '1955' '1957' '1945' '1960' '1967' '1932' '1924' '1964' '2012'
 '1911' '1927' '1948' '1962' '2006' '1952' '1940' '1951' '1931' '1954

In [104]:
import numpy as np

# Функция для преобразования значений в целые числа
def to_integer(year):
    try:
        return int(year)
    except ValueError:
        return np.nan

# Применение функции к столбцу
books['Year-Of-Publication'] = books['Year-Of-Publication'].apply(to_integer).astype('Int64')

# Проверка преобразования типов данных
print(books.dtypes)

ISBN                   object
Book-Title             object
Book-Author            object
Year-Of-Publication     Int64
Publisher              object
Image-URL-S            object
Image-URL-M            object
Image-URL-L            object
dtype: object


2. **Проведем первичный анализ**

In [105]:
# Первичный анализ
print(books.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 8 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   ISBN                 271360 non-null  object
 1   Book-Title           271360 non-null  object
 2   Book-Author          271358 non-null  object
 3   Year-Of-Publication  271357 non-null  Int64 
 4   Publisher            271358 non-null  object
 5   Image-URL-S          271360 non-null  object
 6   Image-URL-M          271360 non-null  object
 7   Image-URL-L          271357 non-null  object
dtypes: Int64(1), object(7)
memory usage: 16.8+ MB
None


In [106]:
print(books.head())

         ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author  Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                 2002     Oxford University Press   
1  Richard Bruce Wright                 2001       HarperFlamingo Canada   
2          Carlo D'Este                 1991             HarperPerennial   
3      Gina Bari Kolata                 1999        Farrar Straus Giroux   
4       E. J. W. Barber                 1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazo

In [107]:
print(ratings.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1149780 entries, 0 to 1149779
Data columns (total 3 columns):
 #   Column       Non-Null Count    Dtype 
---  ------       --------------    ----- 
 0   User-ID      1149780 non-null  int64 
 1   ISBN         1149780 non-null  object
 2   Book-Rating  1149780 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 26.3+ MB
None


In [108]:
print(ratings.head())

   User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6


In [109]:
print(users.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278858 entries, 0 to 278857
Data columns (total 3 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   User-ID   278858 non-null  int64  
 1   Location  278858 non-null  object 
 2   Age       168096 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 6.4+ MB
None


In [110]:
print(users.head())

   User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN


### 2: Очистка данных

**Проверка и обработка пропущенных значений**:
   - Проверим наличие пропущенных значений и решим, как с ними поступить (удалить или заполнить).


In [111]:
# Проверка пропущенных значений
print(books.isnull().sum())

ISBN                   0
Book-Title             0
Book-Author            2
Year-Of-Publication    3
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64


In [112]:
print(ratings.isnull().sum())

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64


In [113]:
print(users.isnull().sum())

User-ID          0
Location         0
Age         110762
dtype: int64


В данном случае, столбцы с URL изображений можно удалить, так как они не имеют значения для анализа.

In [114]:
books = books.drop(columns=['Image-URL-S', 'Image-URL-M', 'Image-URL-L'])

Для столбцов с пропущенными значениями (`Book-Author`, `Publisher`, `Year-Of-Publication`), можно воспользоваться различными подходами в зависимости от специфики нашего анализа.

In [115]:
# Заполним пропуски значением "Unknown"
books['Book-Author'] = books['Book-Author'].fillna('Unknown')
books['Publisher'] = books['Publisher'].fillna('Unknown')

Для `Year-Of-Publication` можно либо заполнить средним значением, либо медианой, либо наиболее часто встречающимся значением.

In [116]:
# Найдем моду (наиболее часто встречающееся значение) для столбца Year-Of-Publication
mode_year = books['Year-Of-Publication'].mode()[0]

# Заполним пропуски модой
books['Year-Of-Publication'] = books['Year-Of-Publication'].fillna(mode_year)

In [117]:
print(books.isnull().sum())

ISBN                   0
Book-Title             0
Book-Author            0
Year-Of-Publication    0
Publisher              0
dtype: int64


Посмотрим на описательную статистику `users`

In [118]:
# Описательная статистика
users.describe()

User-ID            Age
count  278858.00000  168096.000000
mean   139429.50000      34.751434
std     80499.51502      14.428097
min         1.00000       0.000000
25%     69715.25000      24.000000
50%    139429.50000      32.000000
75%    209143.75000      44.000000
max    278858.00000     244.000000

In [119]:
# Найдем значения возраста больше 100
ages_above_100 = users[users['Age'] > 100]['Age']

# Выведем эти значения
print(ages_above_100)

1288      103.0
1322      104.0
1578      231.0
3084      104.0
3210      119.0
          ...  
276047    127.0
276352    104.0
277107    104.0
277503    103.0
278301    104.0
Name: Age, Length: 366, dtype: float64


Заменим `ages_above_100`, а также пропущенные значения (NaN), на медиану возраста.

Кажется, что возраст больше 100 лет – это аномальные значения (выбросы). Медиана возраста составляет 32 года. Мы можем использовать эту медиану для замены выбросов и пропущенных значений.

In [120]:
# Найдем медиану возраста
age_median = users['Age'].median()

# Заменим выбросы (возраст > 100) и пропущенные значения на медиану
users['Age'] = users['Age'].apply(lambda x: age_median if x > 100 or pd.isnull(x) else x)

# Проверим результат
print(users['Age'].describe())

count    278858.000000
mean         33.544844
std          10.789130
min           0.000000
25%          29.000000
50%          32.000000
75%          35.000000
max         100.000000
Name: Age, dtype: float64


In [121]:
print(users.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278858 entries, 0 to 278857
Data columns (total 3 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   User-ID   278858 non-null  int64  
 1   Location  278858 non-null  object 
 2   Age       278858 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 6.4+ MB
None


In [122]:
users['Age'] = users['Age'].round(0).astype(int)

In [123]:
print(users.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278858 entries, 0 to 278857
Data columns (total 3 columns):
 #   Column    Non-Null Count   Dtype 
---  ------    --------------   ----- 
 0   User-ID   278858 non-null  int64 
 1   Location  278858 non-null  object
 2   Age       278858 non-null  int32 
dtypes: int32(1), int64(1), object(1)
memory usage: 5.3+ MB
None


**Удаление строк с пропусками:**<br>
Обработка пропущенных значений (например, удаление строк с пропущенными значениями)

In [124]:
books = books.dropna()
ratings = ratings.dropna()
users = users.dropna()

**Удаление дубликатов:**<br>
Проверим наличие дубликатов и удалите их.

In [125]:
# Проверка дубликатов
print(books.duplicated().sum())
print(ratings.duplicated().sum())
print(users.duplicated().sum())

# Удаление дубликатов
books = books.drop_duplicates()
ratings = ratings.drop_duplicates()
users = users.drop_duplicates()

0
0
0


### 3: Подготовка данных

**Объединение данных:**
    - Объединим данные из разных файлов на основе общих столбцов (например, `book_id` и `user_id`).

In [126]:
print(books.info())
print(ratings.info())
print(users.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 5 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   ISBN                 271360 non-null  object
 1   Book-Title           271360 non-null  object
 2   Book-Author          271360 non-null  object
 3   Year-Of-Publication  271360 non-null  Int64 
 4   Publisher            271360 non-null  object
dtypes: Int64(1), object(4)
memory usage: 10.6+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1149780 entries, 0 to 1149779
Data columns (total 3 columns):
 #   Column       Non-Null Count    Dtype 
---  ------       --------------    ----- 
 0   User-ID      1149780 non-null  int64 
 1   ISBN         1149780 non-null  object
 2   Book-Rating  1149780 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 26.3+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278858 entries, 0 to 278857
Data columns (total 3 c

В нашем случае,  у нас есть три датафрейма **`books`**, **`ratings`** и **`users`**, и они имеют общие столбцы `ISBN` и `User-ID`.

In [127]:
# Использование внутреннего соединения (inner join)
final_data = pd.merge(pd.merge(books, ratings, on='ISBN'), users, on='User-ID', how='inner')

In [128]:
print(final_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1031136 entries, 0 to 1031135
Data columns (total 9 columns):
 #   Column               Non-Null Count    Dtype 
---  ------               --------------    ----- 
 0   ISBN                 1031136 non-null  object
 1   Book-Title           1031136 non-null  object
 2   Book-Author          1031136 non-null  object
 3   Year-Of-Publication  1031136 non-null  Int64 
 4   Publisher            1031136 non-null  object
 5   User-ID              1031136 non-null  int64 
 6   Book-Rating          1031136 non-null  int64 
 7   Location             1031136 non-null  object
 8   Age                  1031136 non-null  int32 
dtypes: Int64(1), int32(1), int64(2), object(5)
memory usage: 67.9+ MB
None


Сохраним и запишем полученный датафрейм в CSV-файл

In [129]:
final_data.to_csv('final_data_labeled.csv', index=False)

## **Задание 2.**

**Выполните разметку на основе правил (rule-based labeling) на подмножестве выбранного датасета.<br>
Разработайте и реализуйте набор правил или условий, которые позволят автоматически присваивать метки данным на основе определенных шаблонов или критериев.**

Для выполнения разметки данных на основе правил (rule-based labeling) в нашем датасете `final_data`, можно разработать и реализовать набор условий, который позволит автоматически присваивать метки данным. 
Например, предположим, что мы хотим разметить данные по следующим критериям:

1. На основе года публикации книги.
2. На основе рейтинга книги.
3. На основе возраста пользователя.
4. На основе количества страниц книги
5. На основе жанра книги
6. На основе количества рецензий

Для добавления новых столбцов в DataFrame `final_data`, можно использовать фиктивные данные в новый DataFrame под именем `final_data_fictitious`, а затем применяем правила для создания меток:

In [130]:
# Создание фиктивного DataFrame
np.random.seed(42)  # Для воспроизводимости результатов

n = 100  # Количество строк в DataFrame

# Генерация случайных данных для DataFrame
data = {
    'Year-Of-Publication': np.random.randint(1950, 2023, size=n),  # Год публикации книги
    'Book-Rating': np.random.randint(1, 11, size=n),  # Рейтинг книги
    'Age': np.random.randint(10, 80, size=n),  # Возраст пользователя
    'Page-Count': np.random.randint(50, 1000, size=n),  # Количество страниц в книге
    'Genre': np.random.choice(['Science Fiction', 'Fantasy', 'Mystery', 'Thriller', 'Romance', 'Drama', 'Non-Fiction'], size=n),  # Жанр книги
    'Review-Count': np.random.randint(0, 100, size=n)  # Количество отзывов на книгу
}

# Создание DataFrame из сгенерированных данных
final_data_fictitious = pd.DataFrame(data)

# Функции для создания меток на основе правил

def label_year_of_publication(row):
    # Метка для года публикации
    if row['Year-Of-Publication'] < 2000:
        return 'Before 2000'
    else:
        return 'After 2000'

def label_book_rating(row):
    # Метка для рейтинга книги
    if row['Book-Rating'] >= 7:
        return 'High'
    elif row['Book-Rating'] >= 4:
        return 'Medium'
    else:
        return 'Low'

def label_user_age(row):
    # Метка для возраста пользователя
    if row['Age'] < 18:
        return 'Child'
    elif row['Age'] < 65:
        return 'Adult'
    else:
        return 'Senior'

def label_page_count(row):
    # Метка для количества страниц в книге
    if row['Page-Count'] < 100:
        return 'Short'
    elif row['Page-Count'] >= 100 and row['Page-Count'] < 300:
        return 'Medium'
    else:
        return 'Long'

def label_genre(row):
    # Метка для жанра книги
    if row['Genre'] in ['Science Fiction', 'Fantasy']:
        return 'Speculative Fiction'
    elif row['Genre'] in ['Mystery', 'Thriller']:
        return 'Suspense'
    elif row['Genre'] in ['Romance', 'Drama']:
        return 'Emotional'
    else:
        return 'Other'

def label_review_count(row):
    # Метка для количества отзывов на книгу
    if row['Review-Count'] < 10:
        return 'Few Reviews'
    elif row['Review-Count'] >= 10 and row['Review-Count'] < 50:
        return 'Moderate Reviews'
    else:
        return 'Many Reviews'

# Применение функций для создания меток к данным DataFrame
final_data_fictitious['Year-Label'] = final_data_fictitious.apply(label_year_of_publication, axis=1)
final_data_fictitious['Rating-Label'] = final_data_fictitious.apply(label_book_rating, axis=1)
final_data_fictitious['Age-Label'] = final_data_fictitious.apply(label_user_age, axis=1)
final_data_fictitious['Page-Count-Label'] = final_data_fictitious.apply(label_page_count, axis=1)
final_data_fictitious['Genre-Label'] = final_data_fictitious.apply(label_genre, axis=1)
final_data_fictitious['Review-Count-Label'] = final_data_fictitious.apply(label_review_count, axis=1)

# Вывод первых нескольких строк DataFrame для проверки результатов
print(final_data_fictitious[['Year-Of-Publication', 'Year-Label', 
                             'Book-Rating', 'Rating-Label', 
                             'Age', 'Age-Label',
                             'Page-Count', 'Page-Count-Label',
                             'Genre', 'Genre-Label',
                             'Review-Count', 'Review-Count-Label']].head())

# Сохранение DataFrame в CSV-файл
final_data_fictitious.to_csv('final_data_fictitious.csv', index=False)

   Year-Of-Publication   Year-Label  Book-Rating Rating-Label  Age Age-Label  \
0                 2001   After 2000            8         High   12     Child   
1                 1964  Before 2000            5       Medium   29     Adult   
2                 2021   After 2000            3          Low   68    Senior   
3                 2010   After 2000            8         High   45     Adult   
4                 1970  Before 2000            6       Medium   28     Adult   

   Page-Count Page-Count-Label            Genre          Genre-Label  \
0         921             Long          Romance            Emotional   
1         735             Long            Drama            Emotional   
2         841             Long      Non-Fiction                Other   
3         675             Long      Non-Fiction                Other   
4         337             Long  Science Fiction  Speculative Fiction   

   Review-Count Review-Count-Label  
0            34   Moderate Reviews  
1           

In [131]:
final_data_fictitious.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Year-Of-Publication  100 non-null    int32 
 1   Book-Rating          100 non-null    int32 
 2   Age                  100 non-null    int32 
 3   Page-Count           100 non-null    int32 
 4   Genre                100 non-null    object
 5   Review-Count         100 non-null    int32 
 6   Year-Label           100 non-null    object
 7   Rating-Label         100 non-null    object
 8   Age-Label            100 non-null    object
 9   Page-Count-Label     100 non-null    object
 10  Genre-Label          100 non-null    object
 11  Review-Count-Label   100 non-null    object
dtypes: int32(5), object(7)
memory usage: 7.5+ KB


## **Задача 3.**

**Выполните разметку вручную отдельного подмножества выбранного датасета с помощью выбранного вами инструмента разметки.**

### 1: Установка и Настройка Label Studio

- Установим *Label Studio* в терминале для нашего виртуального окружения:

In [132]:
# pip install label-studio

- Настроим проект в Label Studio:
    - Создадим новый проект.
    - Укажем типы разметки и настроим интерфейсы разметки в зависимости от структуры данных. Для текстовых данных можно настроить разметку категорий.

### 2: Подготовка Подмножества Данных для Разметки

Извлечем подмножество данных из `final_data_fictitious`для ручной разметки:

In [133]:
# Предположим, что мы размечаем 20% данных вручную
split_ratio = 0.2
subset_size = int(len(final_data_fictitious) * split_ratio)
manual_subset_df = final_data_fictitious.sample(n=subset_size, random_state=42)

# Сохраним подмножество данных в формате CSV для загрузки в Label Studio
manual_subset_df.to_csv('manual_subset.csv', index=False)

### 3: Импорт Данных в Label Studio и Разметка

1. Загрузим `manual_subset.csv`в созданный проект *Label Studio.*

2. Проведем ручную разметку данных.

In [134]:
# Запускаем в терминале Label Studio.
# label-studio

1. Создаем новый проект `Book_Recommendation` (Проведение разметки вручную отдельного подмножества датасета `final_data_fictitious`/Manually marking up a separate subset of the final_data_fictitious dataset)
2. Загрузим `manual_subset.csv` в созданный проект *Label Studio*.
3. Проведем ручную разметку данных.
    - Создаем конфигурацию меток:
        - Добавим метки для каждой категории данных. (Для корректной разметки данных, нам нужно создать отдельные группы выбора для каждой категории)
4. Назначение меток:
    - После настройки меток, делаем разметку данных вручную.
5.  Проверка и экспорт разметки:
    - Проверим несколько записей, чтобы убедиться, что метки назначены правильно.
    - После завершения разметки экспортируем данные с метками для дальнейшего анализа или использования в моделях машинного обучения.

### 4: Экспорт Размеченных Данных из Label Studio

После завершения разметки экспортируем размеченные данные из *Label Studio* в формате CSV или JSON и загрузим их обратно в Python:

In [135]:
# Экспортированный файл - 'annotation_result.json'
manual_annotations = pd.read_json('annotation_result.json')

## **Задача 4.**

**Объедините данные, размеченные вручную, с данными, размеченными на основе правил.<br>
Объедините два подмножества размеченных данных в один набор данных, сохранив при этом соответствующую структуру и целостность.**

In [136]:
# Подгрузим два набора данных
final_data_fictitious = pd.read_csv('final_data_fictitious.csv')
manual_annotations = pd.read_csv('annotation_result.csv')

# Объединение данных
combined_data = pd.concat([final_data_fictitious, manual_annotations], ignore_index=True)

In [137]:
combined_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120 entries, 0 to 119
Data columns (total 24 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Year-Of-Publication  120 non-null    int64  
 1   Book-Rating          120 non-null    int64  
 2   Age                  120 non-null    int64  
 3   Page-Count           120 non-null    int64  
 4   Genre                120 non-null    object 
 5   Review-Count         120 non-null    int64  
 6   Year-Label           120 non-null    object 
 7   Rating-Label         120 non-null    object 
 8   Age-Label            120 non-null    object 
 9   Page-Count-Label     120 non-null    object 
 10  Genre-Label          120 non-null    object 
 11  Review-Count-Label   120 non-null    object 
 12  age                  20 non-null     object 
 13  annotation_id        20 non-null     float64
 14  annotator            20 non-null     float64
 15  created_at           20 non-null     obj

In [138]:
combined_data.head(10)

Year-Of-Publication  Book-Rating  Age  Page-Count            Genre  \
0                 2001            8   12         921          Romance   
1                 1964            5   29         735            Drama   
2                 2021            3   68         841      Non-Fiction   
3                 2010            8   45         675      Non-Fiction   
4                 1970            6   28         337  Science Fiction   
5                 1973            3   76         992         Thriller   
6                 1952            1   28         903          Romance   
7                 1971            3   29         712  Science Fiction   
8                 2002            5   61         688          Fantasy   
9                 1951            3   42         204          Fantasy   

   Review-Count   Year-Label Rating-Label Age-Label Page-Count-Label  ...  \
0            34   After 2000         High     Child             Long  ...   
1            86  Before 2000       Medium     Adult             Long  ...   
2            80   After 2000          Low    Senior             Long  ...   
3            89   After 2000         High     Adult             Long  ...   
4             7  Before 2000       Medium     Adult             Long  ...   
5            92  Before 2000          Low    Senior             Long  ...   
6            25  Before 2000          Low     Adult             Long  ...   
7            73  Before 2000          Low     Adult             Long  ...   
8            89   After 2000       Medium     Adult             Long  ...   
9            33  Before 2000          Low     Adult           Medium  ...   

  annotator created_at genre  id  lead_time page_count rating  review_count  \
0       NaN        NaN   NaN NaN        NaN        NaN    NaN           NaN   
1       NaN        NaN   NaN NaN        NaN        NaN    NaN           NaN   
2       NaN        NaN   NaN NaN        NaN        NaN    NaN           NaN   
3       NaN        NaN   NaN NaN        NaN        NaN    NaN           NaN   
4       NaN        NaN   NaN NaN        NaN        NaN    NaN           NaN   
5       NaN        NaN   NaN NaN        NaN        NaN    NaN           NaN   
6       NaN        NaN   NaN NaN        NaN        NaN    NaN           NaN   
7       NaN        NaN   NaN NaN        NaN        NaN    NaN           NaN   
8       NaN        NaN   NaN NaN        NaN        NaN    NaN           NaN   
9       NaN        NaN   NaN NaN        NaN        NaN    NaN           NaN   

   updated_at year  
0         NaN  NaN  
1         NaN  NaN  
2         NaN  NaN  
3         NaN  NaN  
4         NaN  NaN  
5         NaN  NaN  
6         NaN  NaN  
7         NaN  NaN  
8         NaN  NaN  
9         NaN  NaN  

[10 rows x 24 columns]

In [139]:
# Сохранение DataFrame в CSV-файл
combined_data.to_csv('combined_data.csv', index=False)

## **Задача 5.**

**Обучите модель машинного обучения, используя объединенный набор размеченных данных.<br>
Разделите датасет на обучающий и тестовый наборы и используйте обучающий набор для обучения модели.**

Разделим объединенные данные на обучающий и тестовый наборы, а затем обучим модель.

In [140]:
from sklearn.model_selection import train_test_split  # используется для разделения данных на обучающую и тестовую выборки
from sklearn.preprocessing import LabelEncoder, StandardScaler  # используется (LabelEncoder) для преобразования категориальных признаков в числовые, (StandardScaler) для нормализации данных.
from sklearn.ensemble import RandomForestClassifier  # используется для создания модели случайного леса.
from sklearn.linear_model import LogisticRegression  # используется для создания модели логистической регрессии.
from sklearn.metrics import accuracy_score  # используется для оценки точности моделей.

Попробовать различные модели, такие как логистическая регрессия и случайный лес, является хорошей практикой.<br>
Это позволяет вам сравнить их производительность и выбрать наиболее подходящую модель для нашей задачи.<br> 
Логистическая регрессия может быть особенно полезной для интерпретации результатов и понимания влияния различных признаков на предсказания.

In [141]:
# Загрузка данных из CSV файла
combined_data = pd.read_csv('combined_data.csv')

# Заполнение отсутствующих данных
# Используем метод forward-fill (ffill) и backward-fill (bfill) для заполнения пропущенных значений.
# ffill заполняет пропуски предыдущими значениями, а bfill - последующими.
combined_data = combined_data.ffill().bfill()

# Преобразование категориальных данных в числовые с помощью LabelEncoder
# LabelEncoder преобразует категориальные текстовые данные в числовые значения.
label_encoders = {}
for column in combined_data.select_dtypes(include=['object']).columns:
    le = LabelEncoder()
    combined_data[column] = le.fit_transform(combined_data[column])
    label_encoders[column] = le

# Разделение датасета на признаки (X) и целевую переменную (y)
# Здесь мы выделяем признаки (все столбцы кроме 'Book-Rating') и целевую переменную ('Book-Rating').
X = combined_data.drop('Book-Rating', axis=1)
y = combined_data['Book-Rating']

# Разделение данных на обучающую и тестовую выборки
# train_test_split делит данные на обучающую и тестовую выборки. test_size=0.2 означает, что 20% данных пойдут в тестовую выборку.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Нормализация признаков с помощью StandardScaler
# Стандартизация данных: приведение всех признаков к одному масштабу со средним значением 0 и стандартным отклонением 1.
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

Обучение модели методами `RandomForestClassifier` и `LogisticRegression`

In [142]:
# Обучение модели RandomForestClassifier
# RandomForestClassifier - это ансамблевый метод, который строит множество деревьев решений и объединяет их результаты.
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [143]:
# Обучение модели LogisticRegression
# LogisticRegression - это метод классификации, который моделирует вероятность принадлежности к одному из классов.
logistic_model = LogisticRegression(random_state=42, max_iter=1000)
logistic_model.fit(X_train, y_train)


LogisticRegression(max_iter=1000, random_state=42)

## **Задача 6.**

**Оценить эффективность обученной модели на тестовом датасете. Используйте подходящие метрики оценки.<br> 
Интерпретируйте результаты и проанализируйте эффективность модели в решении задачи разметки.**

In [144]:
# Предсказание значений на тестовой выборке с использованием RandomForestClassifier
# Используем обученную модель для предсказания классов на тестовых данных.
y_pred_rf = rf_model.predict(X_test)

# Оценка точности модели RandomForestClassifier с использованием метрики accuracy_score
# accuracy_score вычисляет долю правильных предсказаний.
accuracy_rf = accuracy_score(y_test, y_pred_rf)
print(f'Random Forest Accuracy: {accuracy_rf}')


# Предсказание значений на тестовой выборке с использованием LogisticRegression
# Используем обученную модель для предсказания классов на тестовых данных.
y_pred_logistic = logistic_model.predict(X_test)

# Оценка точности модели LogisticRegression с использованием метрики accuracy_score
accuracy_logistic = accuracy_score(y_test, y_pred_logistic)
print(f'Logistic Regression Accuracy: {accuracy_logistic}')

Random Forest Accuracy: 0.5416666666666666
Logistic Regression Accuracy: 0.4166666666666667


- В данном случае, модель Random Forest имеет точность примерно 54.2%.
- В данном случае, модель логистической регрессии имеет точность примерно 41.7%.

#### Интерпретация результатов:

- **Сравнение моделей**: На основании этих результатов можно сказать, что модель Random Forest показывает лучшую производительность на данном наборе данных по сравнению с моделью Logistic Regression.
- **Абсолютные значения точности**: Однако, обе модели показывают достаточно низкую точность. Это может указывать на то, что либо данные сложные для предсказания, либо модели нуждаются в дополнительной настройке и улучшении, либо есть необходимость проверить качество данных и признаки.

#### Возможные следующие шаги:

1. **Анализ данных**: Провести дополнительный анализ данных для выявления возможных проблем с качеством данных.
2. **Улучшение модели**: Подумать о создании новых признаков, тестировании других моделей или настройке гиперпараметров текущих моделей.
3. **Увеличение данных**: Попробовать собрать дополнительные данные или использовать методы увеличения данных (data augmentation).
4. **Кросс-валидация**: Провести кросс-валидацию для более точного измерения производительности моделей и уменьшения риска переобучения.

Но точность (accuracy) не всегда является лучшей метрикой для оценки модели, особенно если данные разбалансированы.<br> 
Возможно стоит рассмотреть использование других метрик, таких как precision, recall, F1-score и ROC-AUC.